lib/property_constraints.js

```diff
@@ -1,95 +1,116 @@
 var Immutable = require('immutable');
 
-var cardinality = {
-  '=': function(val, count) {
-    return val.size === parseInt(count);
-  },
-  '>': function(val, count) {
-    return val.size > parseInt(count);
-  },
-  '<': function(val, count) {
-    return val.size < parseInt(count);
-  }
+function not(fn) {
+  return function(state) {
+    return !fn(state);
+  };
+}
+
+function and() {
+  var fns = Array.prototype.slice.call(arguments);
+  return function(state) {
+    return fns.reduce(function(memo, fn) {
+      return memo && fn(state);
+    }, true);
+  };
+}
+
+
+var constraint = function() {
+  var fns = Array.prototype.slice.call(arguments);
+  return function(state) {
+    return fns.reduce(function(previousValue, fn) {
+      return (!!previousValue) && fn(previousValue);
+    }, state);
+  };
 };
 
-var handleFunctionParameter = function(expectedSet, propertyValue) {
-  if (typeof (expectedSet) === 'function' && Immutable.Iterable.isIterable(propertyValue)) {
-    return propertyValue.filter(expectedSet);
-  }
-  return expectedSet;
+
+var hasProperty = function(propertyName) {
+  return function(state) {
+    return state.hasOwnProperty(propertyName) && !!state[propertyName];
+  };
 };
 
-var rules = {
+var get = function(attrName) {
+  return function(state) {
+    return state[attrName];
+  };
+};
 
-  hasProperty: function(model, propertyName) {
-    return model.get(propertyName);
-  },
+var getIn = function() {
```

Comment:
This has an interesting consequence for deep structures.

I had sort of made a design decision that 'Immutable' would not be exposed to clients:

Imagine the following scenario:
```javascript
var state = {foo: {bar: 1}};

getIn(['foo', 'bar'])(state) // works fine => 1
getIn(['foo'])(state) // no bueno => Immutable.Map({bar: 1}), should be just {bar: 1}
```

So the coersion should probably be reversed on the way out.